blockchain/ethereum.go

```diff
@@ -536,6 +537,69 @@ func (e *EthereumClient) ProcessEvent(name string, event *types.Log, confirmedCh
 	return nil
 }
 
+// PollFinalizedHeader continuously polls the latest finalized header and stores it in the client
+func (e *EthereumClient) PollFinality() {
+	e.FinalizedHeader.Store(newGlobalFinalizedHeaderManager(e))
+}
+
+// StopPollingForFinalizedHeader stops polling for the latest finalized header
+func (e *EthereumClient) CancelFinalityPolling() {
+	if _, ok := e.headerSubscriptions[finalizedHeaderKey]; ok {
+		e.DeleteHeaderEventSubscription(finalizedHeaderKey)
+	}
+}
+
+// WaitForTxTobeFinalized waits for a transaction to be finalized
+// If the network is simulated, it will return immediately
+// otherwise it waits for the transaction to be finalized and returns the block number and time of the finalization
+func (e *EthereumClient) WaitForFinalizedTx(txHash common.Hash) (*big.Int, time.Time, error) {
+	if e.NetworkSimulated() {
+		return nil, time.Time{}, nil
+	}
+
+	receipt, err := e.Client.TransactionReceipt(context.Background(), txHash)
+	if err != nil {
+		return nil, time.Time{}, err
+	}
+	txHdr, err := e.HeaderByNumber(context.Background(), receipt.BlockNumber)
+	if err != nil {
+		return nil, time.Time{}, err
+	}
+	finalizer := NewTransactionFinalizer(e, txHdr, receipt.TxHash)
+	key := "txFinalizer-" + txHash.String()
+	e.AddHeaderEventSubscription(key, finalizer)
+	defer e.DeleteHeaderEventSubscription(key)
+	err = finalizer.Wait()
+	if err != nil {
+		return nil, time.Time{}, err
+	}
+	return finalizer.FinalizedBy, finalizer.FinalizedAt, nil
+}
+
+// IsTxFinalized checks if the transaction is finalized on chain
+// if the tx is not finalized it returns false, the latest finalized header number and the time at which it was finalized
+// if the tx is finalized it returns true, the finalized header number by which the tx was considered finalized and the time at which it was finalized
+// In case of simulated network, it always returns true
+func (e *EthereumClient) IsTxFinalized(txHdr, header *SafeEVMHeader) (bool, *big.Int, time.Time, error) {
```

Comment:
Naming: you're only comparing headers here, I think it's more apt to call this `IsHeaderFinalized`, and maybe also add `IsTxFinalized` that takes in just a tx hash and checks if that would be useful?

Reply:
To know if a Tx is finalized you will need the block number from its receipt. unless a tx is confirmed it can't be finalized.  From that point `IsTxFinalized` will always mean `IsTxHeaderFinalized`. So we won't need `IsTxFinalized` but I can change the naming to reflect `IsHeaderFinalized`